specs/networking/p2p-interface.md

```diff
@@ -0,0 +1,743 @@
+- `proposer_slashing` - This topic is used solely for propagating proposer slashings to proposers on the network. Proposer slashings are sent in their entirety. Clients who receive a proposer slashing on this topic MUST validate the conditions within `process_proposer_slashing` before forwarding it across the network.
+- `attester_slashing` - This topic is used solely for propagating attester slashings to proposers on the network. Attester slashings are sent in their entirety. Clients who receive an attester slashing on this topic MUST validate the conditions within `process_attester_slashing` before forwarding it across the network.
+
+#### Interop
+
+Unaggregated attestations from all shards are sent to the `beacon_attestation` topic.
+
+#### Mainnet
+
+Shards are grouped into their own subnets (defined by a shard topic). The number of shard subnets is defined via `SHARD_SUBNET_COUNT` and the shard `shard_number % SHARD_SUBNET_COUNT` is assigned to the topic: `shard{shard_number % SHARD_SUBNET_COUNT}_beacon_attestation`. Unaggregated attestations are sent to the subnet topic. Aggregated attestations are sent to the `beacon_attestation` topic.
+
+### Messages
+
+Each gossipsub [message](https://github.com/libp2p/go-libp2p-pubsub/blob/master/pb/rpc.proto#L17-L24) has a maximum size of `GOSSIP_MAX_SIZE`.
+
+Clients MUST reject (fail validation) messages that are over this size limit. Likewise, clients MUST NOT emit or propagate messages larger than this limit.
+
+The payload is carried in the `data` field of a gossipsub message, and varies depending on the topic:
+
+
+| Topic                        | Message Type      |
+|------------------------------|-------------------|
+| beacon_block                 | BeaconBlock       |
+| beacon_attestation           | Attestation       |
+| shard{N}\_beacon_attestation | Attestation       |
+| voluntary_exit               | VoluntaryExit     |
+| proposer_slashing            | ProposerSlashing  |
+| attester_slashing            | AttesterSlashing  |
+
+Clients MUST reject (fail validation) messages containing an incorrect type, or invalid payload.
+
+When processing incoming gossip, clients MAY descore or disconnect peers who fail to observe these constraints.
+
+### Encodings
+
+Topics are post-fixed with an encoding. Encodings define how the payload of a gossipsub message is encoded.
+
+#### Interop
+
+- `ssz` - All objects are SSZ-encoded. Example: The beacon block topic string is: `/beacon_block/ssz` and the data field of a gossipsub message is an ssz-encoded `BeaconBlock`.
+
+#### Mainnet
+
+- `ssz_snappy` - All objects are ssz-encoded and then compressed with snappy. Example: The beacon attestation topic string is: `/beacon_attestation/ssz_snappy` and the data field of a gossipsub message is an `Attestation` that has been ssz-encoded then compressed with snappy.
```

Comment:
Since topics are plain UTF-8 strings, does it make sense to have a map of `code -> encoding`, so that we can save some space in these topics? e.g., `ssz` is `1` and `ssz_snappy` is `2`, so the topics are `/beacon_attestation/1` and `/beacon_attestation/2` .

Reply:
if we do this, we have a higher-value target in `beacon_attestation` - ie one could imagine having a fourcc here where the first three chars are topic and the last is encoding or something like this. that said, if we get stream compression, that should take care of it.

Also, we're sending SSZ - it's not the most bandwidth-friendly format out there already.. not sure this will make a practical difference.

all that said, if there's strong support, I'm fine with indices too.